File: python/templative/lib/distribute/gameCrafter/util/gameCrafterSession.py
import os
import ssl
import aiohttp
from templative.lib.distribute.gameCrafter.util import httpOperations

class GameCrafterSession:
    def __init__(self, httpSession):
        self.httpSession = httpSession
        self.sessionId = None
        self.userId = None

    def login(self, sessionId, userId):
        self.sessionId = sessionId
        self.userId = userId

    async def close(self):
        if self.httpSession and not self.httpSession.closed:
            await self.httpSession.close()

async def login(publicApiKey, userName, userPassword):
    connector = aiohttp.TCPConnector(
        verify_ssl=False,
        enable_cleanup_closed=True
    )
    
    gameCrafterSession = GameCrafterSession(
        aiohttp.ClientSession(
            connector=connector,
            timeout=aiohttp.ClientTimeout(total=30)
        )
    )
    
    try:
        login = await httpOperations.login(gameCrafterSession, publicApiKey, userName, userPassword)
        gameCrafterSession.login(sessionId=login["id"], userId=login["user_id"])
        return gameCrafterSession
    except Exception as e:
        await gameCrafterSession.httpSession.close()
        raise e

async def logout(gameCrafterSession):
    await httpOperations.logout(gameCrafterSession)

async def createSessionFromLogin(id, userId):
    connector = aiohttp.TCPConnector(
        verify_ssl=False,
        enable_cleanup_closed=True
    )
    
    session = aiohttp.ClientSession(
        connector=connector,
        timeout=aiohttp.ClientTimeout(total=30)
    )
    
    gameCrafterSession = GameCrafterSession(session)
    gameCrafterSession.login(sessionId=id, userId=userId)
    return gameCrafterSession

async def logout(gameCrafterSession):
    await httpOperations.logout(gameCrafterSession)

File: src/main/templative/lib/distribute/gameCrafter/util/gameCrafterSession.js
const axios = require('axios');
const { login: gameCrafterLogin, logout: gameCrafterLogout } = require('./httpOperations.js');

class GameCrafterSession {
  constructor() {
    this.axiosInstance = axios.create({
      timeout: 30000, // 30 seconds in milliseconds
      httpsAgent: new (require('https')).Agent({ rejectUnauthorized: false })
    });
    this.sessionId = null;
    this.userId = null;
  }

  login(sessionId, userId) {
    this.sessionId = sessionId;
    this.userId = userId;
  }

  async close() {
    // Axios doesn't require manual closing, but we'll keep the method
    // for interface compatibility
    this.axiosInstance = null;
  }
}

async function login(publicApiKey, userName, userPassword) {
  const gameCrafterSession = new GameCrafterSession();

  try {
    const loginResult = await gameCrafterLogin(gameCrafterSession, publicApiKey, userName, userPassword);
    gameCrafterSession.login(loginResult.id, loginResult.user_id);
    return gameCrafterSession;
  } catch (e) {
    await gameCrafterSession.close();
    throw e;
  }
}

async function logout(gameCrafterSession) {
  await gameCrafterLogout(gameCrafterSession);
}

async function createSessionFromLogin(id, userId) {
  const gameCrafterSession = new GameCrafterSession();
  gameCrafterSession.login(id, userId);
  return gameCrafterSession;
}

module.exports = { GameCrafterSession, login, logout, createSessionFromLogin };
